src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdClientConfiguration.cs

```diff
@@ -0,0 +1,77 @@
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Modules;
+using OrchardCore.OpenId.Services;
+
+namespace OrchardCore.OpenId.Configuration
+{
+    [Feature(OpenIdConstants.Features.Client)]
+    public class OpenIdClientConfiguration : IConfigureOptions<AuthenticationOptions>, IConfigureNamedOptions<OpenIdConnectOptions>
+    {
+        private readonly IOpenIdClientService _openIdConnectService;
+        private readonly ILogger<OpenIdClientConfiguration> _logger;
+
+        public OpenIdClientConfiguration(IOpenIdClientService openIdConnectService, ILogger<OpenIdClientConfiguration> logger)
+        {
+            _openIdConnectService = openIdConnectService;
+            _logger = logger;
+        }
+
+        public void Configure(AuthenticationOptions options)
+        {
+            var settings = _openIdConnectService.GetOpenIdConnectSettings().GetAwaiter().GetResult();
+            if (!_openIdConnectService.IsValidOpenIdConnectSettings(settings))
+            {
+                _logger.LogWarning("The OpenID Connect module is not correctly configured.");
+                return;
+            }
+
+            // Register the OpenIddict handler in the authentication handlers collection.
+            options.AddScheme(OpenIdConnectDefaults.AuthenticationScheme, builder =>
+            {
+                builder.DisplayName = settings.DisplayName;
+                builder.HandlerType = typeof(OpenIdConnectHandler);
+            });
+        }
+
+        public void Configure(string name, OpenIdConnectOptions options)
+        {
+            var settings = _openIdConnectService.GetOpenIdConnectSettings().GetAwaiter().GetResult();
+            if (name != OpenIdConnectDefaults.AuthenticationScheme)
+                return;
+
+            if (_openIdConnectService.IsValidOpenIdConnectSettings(settings))
+            {
+                options.Authority = settings.Authority;
+                options.ClientId = settings.ClientId;
+                options.SignedOutRedirectUri = settings.SignedOutRedirectUri;
+                options.SignedOutCallbackPath = settings.SignedOutCallbackPath;
+
+                options.RequireHttpsMetadata = !settings.TestingModeEnabled;
+
+                options.ResponseType = Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectResponseType.IdToken;
+                options.CallbackPath = settings.CallbackPath;
+
+                foreach (var item in settings?.AllowedScopes)
+                {
+                    options.Scope.Add(item);
+                }
+
+                options.TokenValidationParameters = new Microsoft.IdentityModel.Tokens.TokenValidationParameters
```

Comment:
I'm all for doing that, but you also need to disable the built-in claims mapping feature in this case:

```csharp
options.SecurityTokenValidator = new JwtSecurityTokenHandler
{
    // Disable the built-in JWT claims mapping feature.
    InboundClaimTypeMap = new Dictionary<string, string>()
};
```

This is a change that will need to be made globally, when we move away from the WS-Fed claims everywhere in Orchard. Let's keep that aside for now.